compiler/src/dotty/tools/dotc/typer/Inliner.scala

```diff
@@ -48,18 +56,126 @@ object Inliner {
         (sym.is(AccessFlags) || sym.privateWithin.exists) &&
         !sym.isContainedIn(inlineSym)
 
+      def preTransform(tree: Tree)(implicit ctx: Context): Tree
+
+      def postTransform(tree: Tree)(implicit ctx: Context) = tree match {
+        case Assign(lhs, rhs) if lhs.symbol.name.is(InlineAccessorName) =>
+          cpy.Apply(tree)(useSetter(lhs), rhs :: Nil)
+        case _ =>
+          tree
+      }
 
-      // TODO: Also handle references to non-public types.
-      // This is quite tricky, as such types can appear anywhere, including as parts
-      // of types of other things. For the moment we do nothing and complain
-      // at the implicit expansion site if there's a reference to an inaccessible type.
       override def transform(tree: Tree)(implicit ctx: Context): Tree =
-        super.transform(accessorIfNeeded(tree)) match {
-          case tree1 @ Assign(lhs: RefTree, rhs) if lhs.symbol.name.is(InlineAccessorName) =>
-            cpy.Apply(tree1)(useSetter(lhs), rhs :: Nil)
-          case tree1 =>
-            tree1
-        }
+        postTransform(super.transform(preTransform(tree)))
+    }
+
+    /** Direct approach: place the accessor with the accessed symbol. This has the
+     *  advantage that we can re-use the receiver as is. But it is only
+     *  possible if the receiver is essentially this or an outer this, which is indicated
+     *  by the test that we can find a host for the accessor.
+     */
+    class MakeInlineableDirect(inlineSym: Symbol) extends MakeInlineableMap(inlineSym) {
+      def preTransform(tree: Tree)(implicit ctx: Context): Tree = tree match {
+        case tree: RefTree if needsAccessor(tree.symbol) =>
+          if (tree.symbol.isConstructor) {
+            ctx.error("Implementation restriction: cannot use private constructors in inline methods", tree.pos)
+            tree // TODO: create a proper accessor for the private constructor
+          }
+          else if (AccessProxies.hostForAccessorOf(tree.symbol).exists) useAccessor(tree)
+          else tree
+        case _ =>
+          tree
+      }
+    }
+
+    /** Fallback approach if the direct approach does not work: Place the accessor method
+     *  in the same class as the inlined method, and let it take the receiver as parameter.
+     *  This is tricky, since we have to find a suitable type for the parameter, which might
+     *  require additional type parameters for the inline accessor. An example is in the
+     *  `TestPassing` class in test `run/inline/inlines_1`:
+     *
+     *    class C[T](x: T) {
+     *      private[inlines] def next[U](y: U): (T, U) = (x, y)
```

Comment:
It could also be package private methods in Java classes. Also, in the future we would like to avoid widening `protected` to public, so it would apply to Scala-defined classes as well.